# **STUDENT AI** - DATA CLEANING

## Objectives

Inspect the dataset and solve any issues that might arise from wrong data types, or missing / wrong values

## Inputs

Continues to assess dataset loaded in previous notebook.

## Outputs

Saves the cleaned dataset back to inputs/dataset folder for futher use


---

# Import required libraries

In [3]:
import os
import sys
import pandas as pd
import pickle
sys.path.append('/workspace/student-AI/src')
from data_management import save_plot, save_df, save_analysis
from io import StringIO

print('All Libraries Loaded')

All Libraries Loaded


# Change working directory

### Set the working directory to notebook parent folder
If the output does not match, click **'clear all outputs'** and then **'restart'** the notebook. 
Then run cells from top to bottom.

In [4]:
current_dir = os.getcwd()
os.chdir(os.path.dirname(current_dir))
current_dir = os.getcwd()
print('If correct, Active Directory should read: /workspace/student-AI')
print(f"Active Directory: {current_dir}")

If correct, Active Directory should read: /workspace/student-AI
Active Directory: /workspace/student-AI


Save functions needed to output cell data to file for use in Dashboard

In [45]:
# def save_df(df, filename, directory='outputs/images/plots'):
#     # Create directory if it doesn't exist
#     if not os.path.exists(directory):
#         os.makedirs(directory)

#     # Define the full path
#     filepath = os.path.join(directory, filename)

#     # If a file with the same name already exists, remove it
#     if os.path.isfile(filepath):
#         os.remove(filepath)

#     # Save the DataFrame using pickle
#     with open(filepath, 'wb') as file:
#         pickle.dump(df, file)
        
# def save_analysis(text, filename, directory='outputs/images/plots'):
#     # Create directory if it doesn't exist
#     if not os.path.exists(directory):
#         os.makedirs(directory)
    
#     # Define the full path
#     filepath = os.path.join(directory, filename)
    
#     # If a file with the same name already exists, remove it
#     if os.path.isfile(filepath):
#         os.remove(filepath)
    
#     # Save the text
#     with open(filepath, 'w') as file:
#         file.write(text)

### Load saved dataset

In [5]:
df = pd.read_csv(f"inputs/dataset/Expanded_data_with_more_features.csv")
df.head()

Unnamed: 0  Gender EthnicGroup          ParentEduc     LunchType TestPrep  \
0           0  female         NaN   bachelor's degree      standard     none   
1           1  female     group C        some college      standard      NaN   
2           2  female     group B     master's degree      standard     none   
3           3    male     group A  associate's degree  free/reduced     none   
4           4    male     group C        some college      standard     none   

  ParentMaritalStatus PracticeSport IsFirstChild  NrSiblings TransportMeans  \
0             married     regularly          yes         3.0     school_bus   
1             married     sometimes          yes         0.0            NaN   
2              single     sometimes          yes         4.0     school_bus   
3             married         never           no         1.0            NaN   
4             married     sometimes          yes         0.0     school_bus   

  WklyStudyHours  MathScore  ReadingScore  WritingScore  
0            < 5         71            71            74  
1         5 - 10         69            90            88  
2            < 5         87            93            91  
3         5 - 10         45            56            42  
4         5 - 10         76            78            75

### Drop Unnamed column that pandas created on import

In [6]:
df.drop(columns=['Unnamed: 0'], inplace=True)
df.head()

Gender EthnicGroup          ParentEduc     LunchType TestPrep  \
0  female         NaN   bachelor's degree      standard     none   
1  female     group C        some college      standard      NaN   
2  female     group B     master's degree      standard     none   
3    male     group A  associate's degree  free/reduced     none   
4    male     group C        some college      standard     none   

  ParentMaritalStatus PracticeSport IsFirstChild  NrSiblings TransportMeans  \
0             married     regularly          yes         3.0     school_bus   
1             married     sometimes          yes         0.0            NaN   
2              single     sometimes          yes         4.0     school_bus   
3             married         never           no         1.0            NaN   
4             married     sometimes          yes         0.0     school_bus   

  WklyStudyHours  MathScore  ReadingScore  WritingScore  
0            < 5         71            71            74  
1         5 - 10         69            90            88  
2            < 5         87            93            91  
3         5 - 10         45            56            42  
4         5 - 10         76            78            75

Print sumamry of Dataset and save to disk for use in dashboard

In [7]:
print("Pandas dataset summary:\n")
buffer = StringIO()
df.info(buf=buffer)
info_str = buffer.getvalue()
save_analysis(info_str, 'dataset_summary.txt')
print(info_str)

Pandas dataset summary:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30641 entries, 0 to 30640
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Gender               30641 non-null  object 
 1   EthnicGroup          28801 non-null  object 
 2   ParentEduc           28796 non-null  object 
 3   LunchType            30641 non-null  object 
 4   TestPrep             28811 non-null  object 
 5   ParentMaritalStatus  29451 non-null  object 
 6   PracticeSport        30010 non-null  object 
 7   IsFirstChild         29737 non-null  object 
 8   NrSiblings           29069 non-null  float64
 9   TransportMeans       27507 non-null  object 
 10  WklyStudyHours       29686 non-null  object 
 11  MathScore            30641 non-null  int64  
 12  ReadingScore         30641 non-null  int64  
 13  WritingScore         30641 non-null  int64  
dtypes: float64(1), int64(3), object(10)
memory usage: 3.3+ MB



### ANALYSIS
At first glance, 30000+ data rows is a very robust dataset to be able to train an ML. <br>
There seems to be no reason why the NrSiblings should be a float data type as you cannot have 0.25 of a brother/sister. <br>
I will treat nrSiblings as a **categorical variable** and confert it to an object type before using it in any modelling phases.

### Get more details of dataset:
This function will analyse each column in the dataset and find the unique values for each. If there are more than 10, it assumes a numerical variable and does not list them all...

In [8]:
# capture the output
buffer = StringIO()

# Get unique values
buffer.write("\n** Detailed list of unique values **\n\n")
for column in df.columns:
    unique_values = df[column].unique()
    num_distinct_values = len(unique_values)
    if num_distinct_values > 10:
        buffer.write(f"{column}: many values - {unique_values[:10]}...\n")
    else:
        buffer.write(f"{column}: {num_distinct_values} distinct values - {unique_values}\n")

# display in the Jupyter Notebook cell
print(buffer.getvalue())
save_analysis(buffer.getvalue(), 'dataset_unique_values.txt')


** Detailed list of unique values **

Gender: 2 distinct values - ['female' 'male']
EthnicGroup: 6 distinct values - [nan 'group C' 'group B' 'group A' 'group D' 'group E']
ParentEduc: 7 distinct values - ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school' nan]
LunchType: 2 distinct values - ['standard' 'free/reduced']
TestPrep: 3 distinct values - ['none' nan 'completed']
ParentMaritalStatus: 5 distinct values - ['married' 'single' 'widowed' nan 'divorced']
PracticeSport: 4 distinct values - ['regularly' 'sometimes' 'never' nan]
IsFirstChild: 3 distinct values - ['yes' 'no' nan]
NrSiblings: 9 distinct values - [ 3.  0.  4.  1. nan  2.  5.  7.  6.]
TransportMeans: 3 distinct values - ['school_bus' nan 'private']
WklyStudyHours: 4 distinct values - ['< 5' '5 - 10' '> 10' nan]
MathScore: many values - [71 69 87 45 76 73 85 41 65 37]...
ReadingScore: many values - [71 90 93 56 78 84 43 64 59 54]...
WritingScore: many values - [74 8

### ANALYSIS
The individual values make sense but will need interpreting for future categorical encoders (convert strings to numbers).<br>
Another issue visible at this stage are the 'nan' values in most Variables indicating "not a number" which indicates missing values.<br>
A more detailed check for missing values is needed ...

### Check for missing values

In [49]:
missing_values = df.isnull().sum()
percentage_missing = (missing_values / len(df)) * 100
percentage_missing = percentage_missing.round(1)
missing_data = pd.DataFrame({'Missing_Values': missing_values, 'Percentage_Missing': percentage_missing})
missing_data['Percentage_Missing'] = missing_data['Percentage_Missing'].astype(str) + '%'
print(missing_data)
save_df(missing_data, 'dataset_missing_values.pkl')

                     Missing_Values Percentage_Missing
Gender                            0               0.0%
EthnicGroup                    1840               6.0%
ParentEduc                     1845               6.0%
LunchType                         0               0.0%
TestPrep                       1830               6.0%
ParentMaritalStatus            1190               3.9%
PracticeSport                   631               2.1%
IsFirstChild                    904               3.0%
NrSiblings                     1572               5.1%
TransportMeans                 3134              10.2%
WklyStudyHours                  955               3.1%
MathScore                         0               0.0%
ReadingScore                      0               0.0%
WritingScore                      0               0.0%


### Analysis
Considering the size of the dataset (30000+ rows) soem missing values can be extrapolated (impupted) without significantly affecting the data relationships in the dataset.
Options are either drop (=delete) rows with missing values - which will lose data that is still containined in that row, or 'fill in the blanks' with a logical value.
I will aseess the best option next.

### Assess how many rows would need to be dropped...

In [18]:
total = len(df)
dropped_data = df.dropna()
dropped_data.info()
deleted_rows = total - len(dropped_data)
percent = (1 - len(dropped_data) / total) * 100
percent_rounded = round(percent)
print(f"\n** Dropping missing data will delete {deleted_rows} rows. ({percent_rounded}%) **")

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19243 entries, 2 to 30640
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Gender               19243 non-null  object 
 1   EthnicGroup          19243 non-null  object 
 2   ParentEduc           19243 non-null  object 
 3   LunchType            19243 non-null  object 
 4   TestPrep             19243 non-null  object 
 5   ParentMaritalStatus  19243 non-null  object 
 6   PracticeSport        19243 non-null  object 
 7   IsFirstChild         19243 non-null  object 
 8   NrSiblings           19243 non-null  float64
 9   TransportMeans       19243 non-null  object 
 10  WklyStudyHours       19243 non-null  object 
 11  MathScore            19243 non-null  int64  
 12  ReadingScore         19243 non-null  int64  
 13  WritingScore         19243 non-null  int64  
dtypes: float64(1), int64(3), object(10)
memory usage: 2.2+ MB

** Dropping missing data wi

## Initial Assumptions

* Dropping null values leaves us with 19243 data rows ... I hypothesize that while this seems like enough data to still achieve the business requirement, dropping this many values (37%) will likely induce an imbalance to the data and bias the dataset. Imputing logical values to fill the gaps is more advisable.

## Imputing Missing Values
let's review the missing_data form the full dataset

In [19]:
missing_data

Missing_Values Percentage_Missing
Gender                            0               0.0%
EthnicGroup                    1840               6.0%
ParentEduc                     1845               6.0%
LunchType                         0               0.0%
TestPrep                       1830               6.0%
ParentMaritalStatus            1190               3.9%
PracticeSport                   631               2.1%
IsFirstChild                    904               3.0%
NrSiblings                     1572               5.1%
TransportMeans                 3134              10.2%
WklyStudyHours                  955               3.1%
MathScore                         0               0.0%
ReadingScore                      0               0.0%
WritingScore                      0               0.0%

### Analysis
The highest missing data rate is TransportMeans at 10.2%. The variable is binary (as the unique values are school_bus and private) so imputing the most common value is an option without inducing too much ambiguity.<br>
If the imputed values induce more errors than desired, another option is to drop the column entirely. LunchType has no missing values and I hypothesize that LunchType and TrasnportMeans indicate similar socio-economic status of a given students family -- eg if they need to rely on a school bus, they most likely will also be relying on free school lunches .. this again might have an impact on other support the student might receive extracurricularly - which could have an influence on the stutends performance eventually.

### Imputing Categorical Variables

Categorical variables with missing values are :
* 'EthnicGroup'
* 'TestPrep'
* 'ParentEduc'
* 'ParentMaritalStatus'
* 'IsFirstChild'
* 'PracticeSport'
* 'TransportMeans'
* 'WklyStudyHours'
* ('LunchType' and 'Gender' have no missing values and do not need to be adjusted)

For these categorical variables I will insert the most common value from the dataset (mode) as that will be closest to the actual value probabalistically.

In [20]:
for column in ['EthnicGroup', 'TestPrep', 'ParentEduc', 'TransportMeans', 'ParentMaritalStatus', 'IsFirstChild', 'PracticeSport', 'WklyStudyHours']:
    mode_value = df[column].mode()[0]
    df[column].fillna(mode_value, inplace=True)

### Imputing Numerical Variables

The only numerical variable from the feature set (not counting the scores which have no missing values) is NrSiblings.
Once the missing values have been imputed, I can also change the data type to a more sensical integer rather than float.
The imputed values will be based on the **median** instead of the **mean** as this is less sensitive to outliers since the variable does contain some 'extreme' values of 6 or more siblings.

In [21]:
median_value = df['NrSiblings'].median()
df['NrSiblings'] = df['NrSiblings'].fillna(median_value).astype(int)

### Quick check for remaining missing values and check datatype change and possible duplicate values:

In [22]:
df.isnull().sum()

Gender                 0
EthnicGroup            0
ParentEduc             0
LunchType              0
TestPrep               0
ParentMaritalStatus    0
PracticeSport          0
IsFirstChild           0
NrSiblings             0
TransportMeans         0
WklyStudyHours         0
MathScore              0
ReadingScore           0
WritingScore           0
dtype: int64

### Analysis
Good, there are no more missign values in the categorical feature varuiables, only NrSiblings remains, as the original dataset list it as a numerical float variable. To check again:

In [23]:
df['NrSiblings'].dtype

dtype('int64')

### Convert nrSiblings Variable to categorical by changing dtype to  string

In [24]:
df['NrSiblings'] = df['NrSiblings'].astype('object')
df['NrSiblings'].dtype

dtype('O')

## Manually Adjust Categorical Variables
Adjust values to more sensible categories. For instance binary categories can already be set to 0 or 1 instead of male/female or yes/no.

In [25]:
# Remap values to unproblematic strings or logical values
study_mapping = {
    '< 5': 'Less than 5 hours',
    '5 - 10': 'Between 5-10 hours',
    '> 10': 'More than 10 hours'
}
test_mapping = {
    'none': "not completed",
    'completed': "completed"
}
bus_mapping = {
    'private': 'private',
    'school_bus': 'schoolbus'
}
parentEduc_mapping = {
    "bachelor's degree": 'bachelors',
    "some college": 'college',
    "master's degree": 'masters',
    "associate's degree": 'associates',
    "high school": 'highschool',
    "some high school": 'highschool',
    "bachelor's degree": 'bachelor',
}
lunch_mapping = {
    "free/reduced": 'free',
    "standard": 'standard',
}
# Remove 'group ' from EthnicGroup Column
df['EthnicGroup'] = df['EthnicGroup'].str.replace('group ', '', case=False)

# Adjust values in the column
df['WklyStudyHours'] = df['WklyStudyHours'].map(study_mapping)
df['TestPrep'] = df['TestPrep'].map(test_mapping)
df['LunchType'] = df['LunchType'].map(lunch_mapping)
df['TransportMeans'] = df['TransportMeans'].map(bus_mapping)
df['ParentEduc'] = df['ParentEduc'].map(parentEduc_mapping)

df

Gender EthnicGroup  ParentEduc LunchType       TestPrep  \
0      female           C    bachelor  standard  not completed   
1      female           C     college  standard  not completed   
2      female           B     masters  standard  not completed   
3        male           A  associates      free  not completed   
4        male           C     college  standard  not completed   
...       ...         ...         ...       ...            ...   
30636  female           D  highschool  standard  not completed   
30637    male           E  highschool  standard  not completed   
30638  female           C  highschool      free      completed   
30639  female           D  associates  standard      completed   
30640    male           B     college  standard  not completed   

      ParentMaritalStatus PracticeSport IsFirstChild NrSiblings  \
0                 married     regularly          yes          3   
1                 married     sometimes          yes          0   
2                  single     sometimes          yes          4   
3                 married         never           no          1   
4                 married     sometimes          yes          0   
...                   ...           ...          ...        ...   
30636              single     sometimes           no          2   
30637              single     regularly           no          1   
30638             married     sometimes           no          1   
30639             married     regularly           no          3   
30640             married         never           no          1   

      TransportMeans      WklyStudyHours  MathScore  ReadingScore  \
0          schoolbus   Less than 5 hours         71            71   
1          schoolbus  Between 5-10 hours         69            90   
2          schoolbus   Less than 5 hours         87            93   
3          schoolbus  Between 5-10 hours         45            56   
4          schoolbus  Between 5-10 hours         76            78   
...              ...                 ...        ...           ...   
30636      schoolbus  Between 5-10 hours         59            61   
30637        private  Between 5-10 hours         58            53   
30638        private  Between 5-10 hours         61            70   
30639      schoolbus  Between 5-10 hours         82            90   
30640      schoolbus  Between 5-10 hours         64            60   

       WritingScore  
0                74  
1                88  
2                91  
3                42  
4                75  
...             ...  
30636            65  
30637            51  
30638            67  
30639            93  
30640            58  

[30641 rows x 14 columns]

### The data is now clean and logical values have been added, in the next book I will conduct an EDA to go into detail about the feature set and data distribution / balance. 

## Save file to repository for follow on notebooks

In [26]:
file_path = 'outputs/dataset/Expanded_data_with_more_features_clean.csv'

# Remove previous file if it exists
if os.path.exists(file_path):
    os.remove(file_path)

# Create the directory if it doesn't exist
os.makedirs(name='outputs/dataset', exist_ok=True)

# Save cleaned DataFrame to the file path
df.to_csv(file_path, index=False)

## Create a test student list random sample
This section will create a csv file of a random sample from the list and remove their scores. This will allow testing the report function in the dashboard where a list of students can be assessed in bulk.

Remove scores

In [28]:
df.drop(columns=['MathScore','ReadingScore','WritingScore'], inplace= True)
df

Gender EthnicGroup  ParentEduc LunchType       TestPrep  \
0      female           C    bachelor  standard  not completed   
1      female           C     college  standard  not completed   
2      female           B     masters  standard  not completed   
3        male           A  associates      free  not completed   
4        male           C     college  standard  not completed   
...       ...         ...         ...       ...            ...   
30636  female           D  highschool  standard  not completed   
30637    male           E  highschool  standard  not completed   
30638  female           C  highschool      free      completed   
30639  female           D  associates  standard      completed   
30640    male           B     college  standard  not completed   

      ParentMaritalStatus PracticeSport IsFirstChild NrSiblings  \
0                 married     regularly          yes          3   
1                 married     sometimes          yes          0   
2                  single     sometimes          yes          4   
3                 married         never           no          1   
4                 married     sometimes          yes          0   
...                   ...           ...          ...        ...   
30636              single     sometimes           no          2   
30637              single     regularly           no          1   
30638             married     sometimes           no          1   
30639             married     regularly           no          3   
30640             married         never           no          1   

      TransportMeans      WklyStudyHours  
0          schoolbus   Less than 5 hours  
1          schoolbus  Between 5-10 hours  
2          schoolbus   Less than 5 hours  
3          schoolbus  Between 5-10 hours  
4          schoolbus  Between 5-10 hours  
...              ...                 ...  
30636      schoolbus  Between 5-10 hours  
30637        private  Between 5-10 hours  
30638        private  Between 5-10 hours  
30639      schoolbus  Between 5-10 hours  
30640      schoolbus  Between 5-10 hours  

[30641 rows x 11 columns]

Optional step<br>
Skip this step if you want to deliberately create an invalid file with missing values. This is to test the report function that it can recognize bad data

In [ ]:
df_clean = df.dropna()

change n=300 to change the size of the student list

In [29]:
df_sample = df_clean.sample(n=300)
df_sample.head()

Gender EthnicGroup  ParentEduc LunchType       TestPrep  \
20371    male           D     college      free  not completed   
7958   female           E     masters  standard  not completed   
25945  female           D     college  standard  not completed   
27299    male           C  highschool  standard  not completed   
30186  female           B  highschool  standard  not completed   

      ParentMaritalStatus PracticeSport IsFirstChild NrSiblings  \
20371              single     regularly          yes          0   
7958              married     regularly          yes          2   
25945              single     regularly          yes          1   
27299            divorced     sometimes           no          1   
30186             married     regularly           no          2   

      TransportMeans      WklyStudyHours  
20371      schoolbus  Between 5-10 hours  
7958       schoolbus   Less than 5 hours  
25945      schoolbus  Between 5-10 hours  
27299      schoolbus   Less than 5 hours  
30186        private   Less than 5 hours

Save the list

In [30]:
file_path = 'inputs/dataset/student_random_list.csv'

# Remove previous file if it exists
if os.path.exists(file_path):
    os.remove(file_path)

# Create the directory if it doesn't exist
os.makedirs(name='inputs/dataset', exist_ok=True)

# Save cleaned DataFrame to the file path
df_sample.to_csv(file_path, index=True)

print("Student List Saved to inputs")

Student List Saved to inputs
